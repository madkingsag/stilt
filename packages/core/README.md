# @stilt/code

*Stilt Framework Conductor*

## Installing Stilt Core

`@stilt/core` on its own does not do much. You need to attach plugins to it.

```javascript
import Stilt from '@stilt/core';

const app = new Stilt();

// install all necessary plugins here.

app.init();
```

## Dependency Injection

Stilt Core comes with built-in dependency injection. Most plugins will provide injectables modules
which you can then inject in your controllers.

Modules will only be injected if the class requesting them is created by the Stilt Core dependency injection mechanism.
Classes that are loaded by one of Stilt's plugins, or are injected will be created using this mechanism.

### Injecting modules

In the following example, we're injecting a module provided by `@stilt/jwt-sessions` inside a `@stilt/rest` controller.

`UserRestController` will be loaded and instantiated by `@stilt/rest` automatically if that plugin is loaded.

```javascript
// user.rest.js

import { Inject } from '@stilt/core';
import { GET } from '@stilt/rest';
import UserModel from './user.model';

@Inject({
    // Access @stilt/jwt-sessions's session manager.
    sessionManager: 'jwt-sessions:session-manager',
})
export default class UserRestController {

    constructor({ sessionManager }) {
        this.sessionManager = sessionManager;
    }

    @GET('/me')
    getLoggedUser() {
        const session = this.sessionManager.getCurrentSession();

        return UserModel.findOne({
            where: { id: session.userId },
        });
    }
}
```

### Creating injectable modules

You can declare your own injectable modules by creating "Injectable Module declaration" files (by default any file ending with `.injectables.js`).
That file should export a default function that returns an Object where the keys are the name of the injectable modules, and the values are the modules themselves.

As an example, let's decouple the API part from the business logic.

```javascript
// user.service.js

import { Inject } from '@stilt/core';
import UserModel from './user.model';

@Inject({
    // Access @stilt/jwt-sessions's session manager.
    sessionManager: 'jwt-sessions:session-manager',
})
export default class UserService {

    constructor({ sessionManager }) {
        this.sessionManager = sessionManager;
    }

    getLoggedUser() {
        const session = this.sessionManager.getCurrentSession();

        return UserModel.findOne({
            where: { id: session.userId },
        });
    }
}
```

```javascript
// user.injectables.js

import UserService from './user.service';

export default function declareInjectables() {

    return {
        'user-service': UserService,
    };
}
```

```javascript
// user.rest.js

@Inject({
    // inject and instanciate user.service.js.
    userService: 'user-service',
})
export default class UserRestController {

    constructor({ userService }) {
        this.userService = userService;
    }

    @GET('/me')
    getLoggedUser() {
        return this.userService.getLoggedUser();
    }
}
```

Note: You can bypass `user.injectables.js` by injecting Classes directly:

```javascript
// user.rest.js

import UserService from './user.service';

@Inject({
    // inject and instanciate user.service.js.
    userService: UserService,
})
export default class UserRestController {

    constructor({ userService }) {
        this.userService = userService;
    }

    @GET('/me')
    getLoggedUser() {
        return this.userService.getLoggedUser();
    }
}
```

### Async Initialization of injectables

There are two ways to initialize injectables asynchronously:

1. Returning a promise resolving to the module, instead of the module:
    ```javascript
    // user.injectables.js

    import UserService from './user.service';

    export default function declareInjectables() {

        return {
            get 'user-service'() {
                return makeSomeAsyncOperation().then(() => UserService);
            },
        };
    }
    ```

2. Tagging a static method of the injectable class with `@AsyncModuleInit`, in which case that method will be called instead of `constructor`:
    ```javascript
    // user.service.js

    import { Inject, AsyncModuleInit } from '@stilt/core';
    import UserModel from './user.model';

    @Inject({
        // Access @stilt/jwt-sessions's session manager.
        sessionManager: 'jwt-sessions:session-manager',
    })
    export default class UserService {

        constructor({ sessionManager }) {
            this.sessionManager = sessionManager;
        }

        @AsyncModuleInit
        static async asyncInit({ sessionManager }) {

            await makeSomeAsyncOperation();

            // delegate to the constructor.
            return new this({ sessionManager });
        }

        getLoggedUser() {
            const session = this.sessionManager.getCurrentSession();

            return UserModel.findOne({
                where: { id: session.userId },
            });
        }
    }
    ```
