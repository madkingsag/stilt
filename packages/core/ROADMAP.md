# ROADMAP

- Dependency injection
    - `@module('moduleName')`
    - ```javascript
      @inject({
        moduleName: MODULE_CLASS,
      }, moduleName, [moduleName, etc])
      ```
    - Allow controller decorator on instances
        - An instance if then created, using @inject to find the necessary modules
        - With `moduleName`, the system has time to initialize
        - Can inject `config` module
        - CLI can request some modules at will. If it requests a Model module or something that depends on a Model, init the database.
    - CLI should be part of core
