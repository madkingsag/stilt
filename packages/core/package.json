{
  "name": "@stilt/core",
  "version": "0.18.2",
  "description": "Core component of the stilt framework",
  "main": "dist/index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "Guylian Cox <guyliancox@gmail.com>",
  "license": "MIT",
  "dependencies": {
    "@stilt/util": "^0.18.0"
  }
}
