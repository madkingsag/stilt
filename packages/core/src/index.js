// @flow

import util from 'util';
import { asyncGlob } from '@stilt/util';
import DependencyInjector from './dependency-injector';

export { AsyncModuleInit, Inject, Inject as inject, AsyncModuleInit as asyncModuleInit } from './dependency-injector';

export interface Plugin {

  init(app: App): void;
  start(): void | Promise<void>;
  close(): void | Promise<void>;
}

type Config = {
  cwd: ?string,
  logLevel: ?string,
  injectables: ?string,
};

export default class App {

  _dependencyInjector = new DependencyInjector();

  _plugins: Map<Symbol, Plugin> = new Map();
  _injectablesGlob;
  _pluginInitPromises = [];

  constructor(config: Config = {}) {

    asyncGlob.cwd = config.cwd;

    this._defaultLogLevel = config.logLevel || 'info';

    this.logger = this.makeLogger('core', { logLevel: this._defaultLogLevel });
    this._injectablesGlob = config.injectables || '**/*.injectable.js';

    this.registerInjectables(App, this);
  }

  _runPluginMethod(methodName) {
    const promises = [...this._plugins.values()].map(plugin => {
      if (plugin[methodName]) {
        this.logger.debug(`Running ${methodName} on ${getName(plugin)}`);
        return plugin[methodName](this);
      }

      return null;
    });

    return Promise.all(promises);
  }

  use(plugin: Plugin) {
    const moduleIdentifier = plugin.MODULE_IDENTIFIER || plugin.constructor.MODULE_IDENTIFIER;
    if (typeof moduleIdentifier !== 'symbol') {
      throw new TypeError(`Trying to load stilt plugin named "${getName(plugin)}" but we cannot uniquely identify it. It is missing the property "MODULE_IDENTIFIER" (must be a Symbol).`);
    }

    this._plugins.set(moduleIdentifier, plugin);

    this._pluginInitPromises.push(plugin.init(this));

    this.registerInjectables(moduleIdentifier, plugin);
    this.registerInjectables(plugin.constructor, plugin);
  }

  getPlugin(pluginIdentifier) {
    if (pluginIdentifier.MODULE_IDENTIFIER) {
      return this.getPlugin(pluginIdentifier.MODULE_IDENTIFIER);
    }

    if (!this._plugins.has(pluginIdentifier)) {
      throw new TypeError(`Plugin with identifier ${String(pluginIdentifier)} has not been registered. Did you forget to register it?`);
    }

    return this._plugins.get(pluginIdentifier);
  }

  makeLogger(namespace: string) {
    // TODO use actual logger & namespace it.
    return console;
  }

  /**
   * Starts the application.
   */
  async start() {
    await Promise.all(this._pluginInitPromises);
    await this._findInjectables();
    await this._runPluginMethod('start');
  }

  /**
   * Closes the application
   */
  async close() {
    await this._runPluginMethod('close');
  }

  async _findInjectables() {
    const injectableFiles = await asyncGlob(this._injectablesGlob);

    const declarations = injectableFiles.map(file => {
      const module = require(file);
      if (module.default) {
        return module.default();
      }

      return module();
    });

    for (const declaration of declarations) {
      this._dependencyInjector.registerAll(declaration);
    }
  }

  instanciate(Class) {
    return this.instantiate(Class);
  }

  instantiate(Class) {
    return this._dependencyInjector.getInstance(Class);
  }

  registerInjectables(map) {
    return this._dependencyInjector.registerAll(map);
  }

  static async createApp(InitModule: Function, config: ?Config) {
    const app = new App(config);
    await app._findInjectables();

    return app.instantiate(InitModule);
  }
}

function getName(obj) {
  return obj && obj.constructor && obj.constructor.name || String(obj);
}

App.prototype.instanciate = util.deprecate(App.prototype.instanciate, 'StiltCore#instanciate is deprecated and renamed into StiltCore#instantiate');
