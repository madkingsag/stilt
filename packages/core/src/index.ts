import {
  asyncGlob,
} from '@stilt/util';
import Emittery from 'emittery';
import createAnnotation, { getPropertyAnnotation } from './annotations.js';
import type { TInstantiable } from './dependency-injector';
import DependencyInjector from './dependency-injector.js';
import type { Factory } from './factory';
import type { TOptionalLazy } from './lazy';
import type { TRunnable } from './runnables';
import type { Class, InjectableIdentifier } from './typing';

export {
  AsyncModuleInit,
  Inject,
  Inject as inject,
  AsyncModuleInit as asyncModuleInit,
} from './dependency-injector.js';
export { runnable, TRunnable, isRunnable } from './runnables.js';
export { factory, Factory, isFactory } from './factory.js';
export { InjectableIdentifier } from './typing.js';
export { lazy } from './lazy.js';

export const AppEvents = Object.freeze({
  Start: createAnnotation('Start', ['method-instance']),
  Close: createAnnotation('Close', ['method-instance']),
});

export type Logger = Console;

type Config = {
  cwd?: string,
  logLevel?: string,
};

enum LifecycleEvents {
  Start = 'start',
  Close = 'close',
}

export class App {

  static LifecycleEvents = LifecycleEvents;

  public readonly lifecycle = new Emittery();

  private readonly _dependencyInjector = new DependencyInjector(newDependency => {
    this._onDependencyInstanciation(newDependency);
  });

  // private _injectablesGlob;

  constructor(config: Config = {}) {
    asyncGlob.cwd = config.cwd;

    // const defaultLogLevel = config.logLevel || 'info';
    // this.logger = this.makeLogger('core', { logLevel: defaultLogLevel });

    // make StiltApp injectable
    this._dependencyInjector.registerInstance(App, this);
  }

  /**
   * Register @AppEvents.Start & @AppEvents.Close listeners
   */
  private _onDependencyInstanciation(dependency: object) {
    const startListeners = Object.keys(getPropertyAnnotation(dependency, AppEvents.Start));
    for (const listenerKey of startListeners) {
      this.lifecycle.on(LifecycleEvents.Start, () => dependency[listenerKey]());
    }

    const closeListeners = Object.keys(getPropertyAnnotation(dependency, AppEvents.Close));
    for (const listenerKey of closeListeners) {
      this.lifecycle.on(LifecycleEvents.Close, () => dependency[listenerKey]());
    }
  }

  makeLogger(_namespace: string, _config?: any) {
    // TODO use actual logger & namespace it.
    return console;
  }

  /**
   * Starts the application.
   */
  async start() {
    // await this._findInjectables();
    await this.lifecycle.emit(LifecycleEvents.Start);
  }

  /**
   * Closes the application
   */
  async close() {
    await this.lifecycle.emit(LifecycleEvents.Close);
  }

  async executeRunnable<Return>(runnable: TRunnable<Return>): Promise<Return> {
    return this._dependencyInjector.executeRunnable(runnable);
  }

  async use<T>(module: InjectableIdentifier | Factory<T>): Promise<T> {
    return this.instantiate(module);
  }

  register<T>(moduleFactory: Factory<T>): void {
    this._dependencyInjector.registerFactory(moduleFactory);
  }

  registerInstance(identifier, instance) {
    return this._dependencyInjector.registerInstance(identifier, instance);
  }

  instantiate<T>(runnable: TOptionalLazy<TInstantiable<T>>): Promise<T>;
  instantiate<T>(moduleArray: Array<TOptionalLazy<TInstantiable<T>>>): Promise<T[]>;
  instantiate<T>(moduleMap: { [key: string]: TOptionalLazy<TInstantiable<T>> }): Promise<{ [key: string]: T }>;
  async instantiate<T>(moduleFactory: TOptionalLazy<TInstantiable<T>>
    | Array<TOptionalLazy<TInstantiable<T>>>
    | { [key: string]: TOptionalLazy<TInstantiable<T>> }): Promise<T | T[] | { [key: string]: T }> {
    // @ts-expect-error
    return this._dependencyInjector.getInstances(moduleFactory);
  }

  static async createApp(
    InitModule: Class<any>,
    config: Config | null | undefined,
  ) {
    const app = new App(config);

    return app.instantiate(InitModule);
  }
}
