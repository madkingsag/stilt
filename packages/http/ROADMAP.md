# ROADMAP

- Injector decorators:
    - @Request
    - @Response
    - @KoaContext

- StiltHttp.currentRequest
- StiltHttp.currentResponse
- StiltHttp.currentKoaContext

- Request.current
- Response.current
- KoaContext.current

- @Body/@Query (Rest?)
