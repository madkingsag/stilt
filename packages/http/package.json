{
  "name": "@stilt/http",
  "version": "0.7.0",
  "description": "Core component of the stilt framework",
  "main": "dist/index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "Guylian Cox <guyliancox@gmail.com>",
  "license": "MIT",
  "dependencies": {
    "@stilt/util": "^0.2.2",
    "async-hooks-map": "^1.1.0",
    "chalk": "^2.4.1",
    "ip": "^1.1.5",
    "koa": "^2.5.1",
    "koa-better-router": "^2.1.1"
  }
}
