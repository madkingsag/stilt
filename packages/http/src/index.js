// @flow

import Koa from 'koa';
import Router from 'koa-better-router';
import ip from 'ip';
import chalk from 'chalk';

export default class StiltHttp {

  static MODULE_IDENTIFIER = Symbol('@stilt/http');

  _declaredEndpoints = [];

  constructor(config) {
    this.port = config.port;

    this.koa = new Koa();
    this.router = Router().loadMethods();
  }

  initPlugin(app) {
    this.logger = app.makeLogger('http');

    this.koa.use(this.router.middleware());

    // TODO only listen plugin if a route is registered
    this.koa.listen(this.port, () => {
      this._printServerStarted(this.port);
    });
  }

  /**
   * Registers a new route on the HTTP server.
   *
   * @param method The HTTP method that may be used.
   * @param path The path part of the URL on which the route is.
   * @param callback The method handling the route.
   */
  registerRoute(method, path, callback) {
    const asyncCallback = asyncToRestify(callback);

    this.router[method](path, asyncCallback);
  }

  /**
   * Add an endpoint of interest when printing server started log
   */
  declareEndpoint(endpointName, endpointPath) {
    this._declaredEndpoints.push({ name: endpointName, path: endpointPath });
  }

  /**
   * @param port
   * @private
   */
  _printServerStarted(port) {
    chalk.enabled = true;

    const localhost = `http://localhost:${port}`;

    const lines = [
      chalk.bold('Access URLs:'),
      '---',
      ['Localhost', localhost],
      ['LAN', `http://${ip.address()}:${port}`],
      '---',
    ];

    if (this._declaredEndpoints.length > 0) {
      for (const endpoint of this._declaredEndpoints) {
        lines.push([endpoint.name, localhost + endpoint.path]);
      }

      lines.push('---');
    }

    lines.push(chalk.blue(`Press ${chalk.italic('CTRL-C')} to stop`));

    this.logger.info(`Server started ${chalk.green('âœ“')}`);
    this.logger.info();
    this.logger.info(printTable(lines));
    this.logger.info();
  }
}

function printTable(lines) {

  let dividerLength = 0;
  let titleLength = 0;
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    if (Array.isArray(line)) {
      const title = line[0];
      if (titleLength < title.length) {
        titleLength = title.length;
      }
    }
  }

  for (let i = 0; i < lines.length; i++) {

    const line = lines[i];
    if (Array.isArray(line)) {
      const [title, content] = line;
      const alignementSpacing = ' '.repeat(titleLength - title.length);
      lines[i] = `${alignementSpacing}${title}: ${chalk.magenta(content)}`;
    }

    // remove control characters as they are not visible but still increase .length.
    const newLine = clearAinsiColors(lines[i]);
    if (newLine.length > dividerLength) {
      dividerLength = newLine.length;
    }
  }

  const divider = chalk.gray('-'.repeat(dividerLength));
  for (let i = 0; i < lines.length; i++) {
    if (lines[i] === '---') {
      lines[i] = divider;
    }
  }

  return lines.map(line => `\t${line}`).join('\n');
}

// TODO support returned streams (ctx.body = someHTTPStream.on('error', ctx.onerror).pipe(PassThrough()))
// TODO handle errors
function asyncToRestify(asyncFunction) {

  return function RestifiedRoute(ctx) {
    const result = asyncFunction(ctx);

    if (result && result.then) {
      return result.then(body => sendResponse(ctx, body));
    }

    return sendResponse(ctx, result);
  };
}

function sendResponse(ctx, body) {
  ctx.body = body;
}

function clearAinsiColors(str: string): string {
  // source: https://stackoverflow.com/questions/25245716/remove-all-ansi-colors-styles-from-strings

  // eslint-disable-next-line no-control-regex
  return str.replace(/[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g, '');
}
