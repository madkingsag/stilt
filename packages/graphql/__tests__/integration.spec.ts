import assert from 'node:assert/strict';
import crypto from 'node:crypto';
import { App } from '@stilt/core';
import { makeControllerInjector, StiltHttp } from '@stilt/http';
import { StiltJwtSessions } from '@stilt/jwt-sessions';
import getPort from 'get-port';
import { createClient } from 'graphql-ws';
import fetch from 'node-fetch';
import ws from 'ws';
import { StiltGraphQl, OnSubscription, Query } from '../src/index.js';

async function buildSimpleApp(): Promise<[app: App, port: number]> {
  const app = new App();

  const port = await getPort();
  await app.use(StiltHttp.configure({ port }));

  return [app, port];
}

class User {}

const WithViewer = makeControllerInjector<[], { http: StiltHttp }>({
  dependencies: {
    http: StiltHttp,
  },
  async run(_ignore, { http }) {
    const context = http.getCurrentContext();
    assert(context != null, 'context is nullish!');

    return { viewer: new User() };
  },
});

describe('Resolvers', () => {
  it('supports custom controller injectors', async () => {
    class MyResolver {
      @WithViewer(0)
      @Query('comment')
      getComment(params: { viewer: User }) {
        assert(params.viewer != null, 'viewer should not be null when using @Query');

        return { content: 'The comment' };
      }
    }

    const [app, port] = await buildSimpleApp();
    await app.use(StiltGraphQl.configure({
      resolversGlob: false,
      typeDefsGlob: false,
      resolvers: [MyResolver],
      typeDefs: [`
        type Query {
          comment: Comment!
        }

        type Comment {
          content: String!
        }
      `],
    }));

    await app.start();

    const res = await fetch(`http://localhost:${port}/graphql`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        query: `
          query {
            comment { content }
          }
        `,
      }),
    });

    const body = await res.json();

    expect(body).toMatchSnapshot();

    await app.close();
  });
});

describe('Subscriptions', () => {
  it('supports creating a Subscription handler with @OnSubscription', async () => {
    class MyResolver {
      @OnSubscription('newComment')
      async *onSubscriptionToNewComment() {
        yield { content: 'first comment' };
        yield { content: 'second comment' };
      }
    }

    const [app, port] = await buildSimpleApp();
    await app.use(StiltGraphQl.configure({
      resolversGlob: false,
      typeDefsGlob: false,
      resolvers: [MyResolver],
      typeDefs: [`
        type Query {
          dummy: Boolean
        }

        type Comment {
          content: String!
        }

        type Subscription {
          newComment: Comment!
        }
      `],
    }));

    await app.start();

    const results = await collectSubscription(port, undefined, `subscription { newComment { content } }`);
    expect(results).toMatchSnapshot();

    await app.close();
  });

  it('supports custom controller injectors', async () => {
    class MyResolver {
      @WithViewer(0)
      @OnSubscription('newComment')
      async *onSubscriptionToNewComment(params: { viewer: User }) {
        assert(params.viewer != null, 'viewer should not be null when using @OnSubscription');

        yield { content: 'first comment' };
        yield { content: 'second comment' };
      }
    }

    const [app, port] = await buildSimpleApp();
    await app.use(StiltGraphQl.configure({
      resolversGlob: false,
      typeDefsGlob: false,
      resolvers: [MyResolver],
      typeDefs: [`
        type Query {
          dummy: Boolean
        }

        type Comment {
          content: String!
        }

        type Subscription {
          newComment: Comment!
        }
      `],
    }));

    await app.start();

    const results = await collectSubscription(port, undefined, `subscription { newComment { content } }`);
    expect(results).toMatchSnapshot();

    await app.close();
  });

  it('integrates with @stilt/jwt-sessions', async () => {
    const WithSession = makeControllerInjector<[], { jwt: StiltJwtSessions }>({
      dependencies: {
        jwt: StiltJwtSessions,
      },
      async run(_ignore, { jwt }) {
        const session = await jwt.getCurrentSession();

        assert(session != null, 'session is nullish!');
        // @ts-expect-error
        assert(session.sub != null, 'session should not be empty');

        return { session };
      },
    });

    class MyResolver {
      @WithSession(0)
      @OnSubscription('newComment')
      async *onSubscriptionToNewComment() {
        yield { content: 'first comment' };
      }
    }

    const [app, port] = await buildSimpleApp();
    await app.use(StiltJwtSessions.configure({
      secret: 'my-unsafe-secret',
    }));

    await app.use(StiltGraphQl.configure({
      resolversGlob: false,
      typeDefsGlob: false,
      resolvers: [MyResolver],
      typeDefs: [`
        type Query {
          dummy: Boolean
        }

        type Comment {
          content: String!
        }

        type Subscription {
          newComment: Comment!
        }
      `],
    }));

    await app.start();

    const results = await collectSubscription(port, 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiI0NyJ9.xUE5hP52MUMkSC_tZtTShHKitdVqPufywVc7aQMEdNg', `subscription { newComment { content } }`);
    expect(results).toMatchSnapshot();

    await app.close();
  });
});

async function collectSubscription(port: number, authToken: string, query: string) {
  const client = createClient({
    url: `ws://localhost:${port}/graphql`,
    webSocketImpl: ws,
    generateID: () => crypto.randomUUID(),
    connectionParams: { authToken },
  });

  return new Promise((resolve, reject) => {
    const collected = [];

    client.subscribe({ query }, {
      error(error) {
        reject(error);
      },
      complete() {
        resolve(collected);
      },
      next(value) {
        collected.push(value);
      },
    });
  });
}
