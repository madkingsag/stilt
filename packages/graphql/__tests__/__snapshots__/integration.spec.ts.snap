// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolvers supports custom controller injectors 1`] = `
Object {
  "data": Object {
    "comment": Object {
      "content": "The comment",
    },
  },
}
`;

exports[`Subscriptions integrates with @stilt/jwt-sessions 1`] = `
Array [
  Object {
    "data": Object {
      "newComment": Object {
        "content": "first comment",
      },
    },
  },
]
`;

exports[`Subscriptions supports creating a Subscription handler with @OnSubscription 1`] = `
Array [
  Object {
    "data": Object {
      "newComment": Object {
        "content": "first comment",
      },
    },
  },
  Object {
    "data": Object {
      "newComment": Object {
        "content": "second comment",
      },
    },
  },
]
`;

exports[`Subscriptions supports custom controller injectors 1`] = `
Array [
  Object {
    "data": Object {
      "newComment": Object {
        "content": "first comment",
      },
    },
  },
  Object {
    "data": Object {
      "newComment": Object {
        "content": "second comment",
      },
    },
  },
]
`;
